src/python/gudhi/representations/vector_methods.py

```diff
@@ -161,53 +172,27 @@ def transform(self, X):
         Returns:
             numpy array with shape (number of diagrams) x (number of samples = **num_landscapes** x **resolution**): output persistence landscapes.
         """
-        num_diag, Xfit = len(X), []
-        x_values = np.linspace(self.sample_range[0], self.sample_range[1], self.new_resolution)
-        step_x = x_values[1] - x_values[0]
-
-        for i in range(num_diag):
-
-            diagram, num_pts_in_diag = X[i], X[i].shape[0]
 
-            ls = np.zeros([self.num_landscapes, self.new_resolution])
+        Xfit = []
+        x_values = self.im_range
+        for diag in X:
+            midpoints, heights = (diag[:, 0] + diag[:, 1]) / 2., (diag[:, 1] - diag[:, 0]) / 2.
+            tent_functions = np.maximum(heights[None, :] - np.abs(x_values[:, None] - midpoints[None, :]), 0)
+            n_points = diag.shape[0]
+            # Get indices of largest elements: can't take more than n_points - 1 (the last ones are in the right position)
+            argpartition = np.argpartition(-tent_functions, min(self.num_landscapes, n_points-1), axis=1)
+            landscapes = np.take_along_axis(tent_functions, argpartition, axis=1)
+            landscapes = landscapes[:, :min(self.num_landscapes, n_points)].T
```

Comment:
I don't understand the new code. Why use argpartition+take_along_axis instead of just partition? Is there no sorting anymore? I understood the previous version (`landscapes = np.sort(tent_functions, axis=1)[:, -self.num_landscapes:][:, ::-1].T`) better, my previous comment was only that the subarray (`[:, -self.num_landscapes:]`) could be taken before sorting instead of after, to save some work.

Reply:
You are completely right. I did not manage to apply your suggestion immediately, and I thought I noticed a bug in it, hence the overcomplicated version. Anyway, I think the current version is in the spirit of your original proposition.